### PR TITLE
connect to SimpleFSList

### PR DESCRIPTION
**Pull request by songgao**
This is a bare minimum that loads folder list from RPC. I haven't made the waiting etc. yet, but wanted to make sure what I have looks good before moving on.

@chrisnojima I couldn't make `takeEveryPure` to work, so I went with the `takeEvery` approach like in chat.

And here's a screenshot that got myself pretty excited:

screen shot 2018-02-08 at 5 02 08 pm

### CODE REVIEW

**Comment by songgao** [comment]
Hmm it looks like the breadcrumbs doesn't work for subfolders. It just doesn't show at all for e.g. `/keybase/public/songgao/me`. I'll try to figure out why tomorrow.

File: shared/constants/fs.js [hunk 1]
```diff
@@ -1,16 +1,20 @@
 // @flow
 import * as I from 'immutable'
 import * as Types from './types/fs'
+import {Buffer} from 'buffer'
```
Comment (chrisnojima) [inline shared/constants/fs.js]: In order for this to flowtype correctly we need to assume buffer exists globally. This is true in practice due to our polyfilling. so we should just delete this line and it'll work

File: shared/constants/fs.js [hunk 2]
```diff
@@ -23,58 +27,17 @@ export const makeState: I.RecordFactory<Types._State> = I.Record({
...
   ]),
 })
+
+export const makeUUID = () => uuidv1(null, Buffer.alloc(16), 0)
+export const wrapOpID = (actionCreator: Function, payload: any) =>
+  actionCreator({
+    opID: makeUUID(),
+    ...payload,
+  })
+export const dispatchFolderListLoad = (dispatch: Dispatch, path: Types.Path) =>
```
Comment (chrisnojima) [inline shared/constants/fs.js]: we tend to not have any action creators anymore. if you need opIDs created on our side i'd make them in the saga and not during the action creation phase

File: shared/constants/fs.js [hunk 3]
```diff
@@ -23,58 +27,17 @@ export const makeState: I.RecordFactory<Types._State> = I.Record({
...
-      makeFolder({
-        children: I.List(['foobar.team']),
-      }),
-    ],
-    [Types.stringToPath('/keybase/private/foo/foo.priv'), makeFile()],
-    [Types.stringToPath('/keybase/private/foo,bar/foo.bar.priv'), makeFile()],
-    [Types.stringToPath('/keybase/public/foo/foo.pub'), makeFile()],
-    [Types.stringToPath('/keybase/public/bar/bar.pub'), makeFile()],
-    [Types.stringToPath('/keybase/team/foobar/foobar.team'), makeFile()],
+    [Types.stringToPath('/keybase/private'), makeFolder()],
```
Comment (chrisnojima) [inline shared/constants/fs.js]: if we're hooking this up this list should also come from the rpc call

File: shared/fs/container.js [hunk 4]
```diff
@@ -26,7 +26,9 @@ const mapStateToProps = (state: TypedState, {routeProps}: OwnProps) => {
   }
```
Comment (chrisnojima) [inline shared/fs/container.js]: I know this isn't your code but 
1. we can use shorthand if the key and value are the same on line 25
```
path,
```
2. this mapping of items should happen in mergeProps and not mapStaetToProps

File: shared/fs/container.js [hunk 5]
```diff
@@ -26,7 +26,9 @@ const mapStateToProps = (state: TypedState, {routeProps}: OwnProps) => {
...
 }
 
-const mapDispatchToProps = (dispatch: Dispatch) => ({})
+const mapDispatchToProps = (dispatch: Dispatch) => ({
+  _loadFolderList: path => Constants.dispatchFolderListLoad(dispatch, Types.stringToPath(path)),
```
Comment (chrisnojima) [inline shared/fs/container.js]: path here is already a Types.Path and not a string
Also method created in mapDispatchToProps need typing of params as the typing is kinda weak here (visible if you do a flow coverage on this file) so make it
```
(path: Types.Path) => dispatch()...
```

File: shared/reducers/fs.js [hunk 6]
```diff
@@ -9,7 +10,9 @@ export default function(state: Types.State = initialState, action: FSGen.Actions
   switch (action.type) {
     case FSGen.resetStore:
       return initialState
-    case FSGen.increaseCount:
+    case FSGen.folderListLoaded:
+      return state.setIn(['pathItems'], I.merge(state.pathItems, action.payload.pathItems))
```
Comment (chrisnojima) [inline shared/reducers/fs.js]: can this be expressed as a mergeIn

File: shared/actions/fs.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
+// @flow
+import * as Constants from '../constants/fs'
+import * as FsGen from './fs-gen'
+import * as I from 'immutable'
+import * as RPCTypes from '../constants/types/rpc-gen'
+import * as Saga from '../util/saga'
+import * as Types from '../constants/types/fs'
+import {posix as Path} from 'path'
+
+import type {SagaGenerator} from '../constants/types/saga'
```
Comment (chrisnojima) [inline shared/actions/fs.js]: this is available in Saga above

File: shared/actions/fs.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+export const fsPathToRpcPathString = (p: Types.Path): string =>
```
Comment (chrisnojima) [inline shared/actions/fs.js]: this belongs in constants

File: shared/actions/fs.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
+// @flow
+import * as Constants from '../constants/fs'
+import * as FsGen from './fs-gen'
+import * as I from 'immutable'
+import * as RPCTypes from '../constants/types/rpc-gen'
+import * as Saga from '../util/saga'
+import * as Types from '../constants/types/fs'
+import {posix as Path} from 'path'
```
Comment (chrisnojima) [inline shared/actions/fs.js]: I don't think we should rename a thing as a known common module. people will assume Path is really path

File: shared/actions/fs.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+import type {SagaGenerator} from '../constants/types/saga'
+
+export const fsPathToRpcPathString = (p: Types.Path): string =>
+  Types.pathToString(p).substring('/keybase'.length)
+
+function* folderList(action: FsGen.FolderListLoadPayload): SagaGenerator<any, any> {
+  if (action.payload.path === '/keybase') {
+    yield Saga.cancel()
```
Comment (chrisnojima) [inline shared/actions/fs.js]: why do we need to cancel

File: shared/actions/fs.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  })
+
+  const result = yield Saga.call(RPCTypes.SimpleFSSimpleFSReadListRpcPromise, {
+    opID: action.payload.opID,
+  })
+
+  const rootPath = action.payload.path
+  const direntToPathAndPathItem = (d: RPCTypes.Dirent) => [
+    Path.join(Types.pathToString(rootPath), d.name),
+    d.direntType === 1 /* DIR_1 */ ? Constants.makeFolder() : Constants.makeFile(),
```
Comment (chrisnojima) [inline shared/actions/fs.js]: we generate constants for this
```
RPCTypes.simpleFSDirentType.dir
```

File: shared/actions/fs.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+export const fsPathToRpcPathString = (p: Types.Path): string =>
+  Types.pathToString(p).substring('/keybase'.length)
+
+function* folderList(action: FsGen.FolderListLoadPayload): SagaGenerator<any, any> {
+  if (action.payload.path === '/keybase') {
+    yield Saga.cancel()
+    return
+  }
+
+  yield Saga.call(RPCTypes.SimpleFSSimpleFSListRpcPromise, {
```
Comment (chrisnojima) [inline shared/actions/fs.js]: So as i mentioned above just make the opID here and use it, we don't need it in the payload

File: shared/actions/fs.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+  yield Saga.call(RPCTypes.SimpleFSSimpleFSWaitRpcPromise, {
+    opID: action.payload.opID,
+  })
+
+  const result = yield Saga.call(RPCTypes.SimpleFSSimpleFSReadListRpcPromise, {
+    opID: action.payload.opID,
+  })
+
+  const rootPath = action.payload.path
```
Comment (chrisnojima) [inline shared/actions/fs.js]: if you're going to pull this out and name it might as well do it first so you can use the name above

File: shared/constants/fs.js (hunk 3 above)
Comment (songgao) [reply shared/constants/fs.js]: I don't think `SimpleFS` supports listing the `/keybase` root directly for now. It has to be e.g. `/keybase/private`. But I guess we can make that change to support it.

File: shared/reducers/fs.js (hunk 6 above)
Comment (songgao) [reply shared/reducers/fs.js]: I was hoping there was a `mergeIn` but didn't find it in immutable.js . Google FAIL!

File: shared/actions/fs.js (hunk 10 above)
Comment (songgao) [reply shared/actions/fs.js]: Same reason as above; RPC doesn't support listing /keybase but we'll get rid of this after the RPC changes is in.

File: shared/actions/fs.js [hunk 14]
```diff
@@ -5,39 +5,32 @@ import * as I from 'immutable'
...
-    opID: action.payload.opID,
-  })
+  const result = yield Saga.call(RPCTypes.SimpleFSSimpleFSReadListRpcPromise, {opID})
 
-  const rootPath = action.payload.path
   const direntToPathAndPathItem = (d: RPCTypes.Dirent) => [
-    Path.join(Types.pathToString(rootPath), d.name),
-    d.direntType === 1 /* DIR_1 */ ? Constants.makeFolder() : Constants.makeFile(),
+    FsPath.join(Types.pathToString(rootPath), d.name),
+    d.direntType === 1 /* RPCTypes.simpleFSDirentType.dir */ ? Constants.makeFolder() : Constants.makeFile(),
```
Comment (chrisnojima) [inline shared/actions/fs.js]: Why not use the variable??

Comment (songgao) [reply shared/actions/fs.js]: Ah I had that in the first place then I wanted to debug something and changed it back to `1` to narrow down possible causes but forgot to change it back. Thanks for catching it

**Comment by songgao** [comment]
Please take another look @chrisnojima @jzila, now with pending progress tracking!

File: shared/actions/fs.js [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+import * as Constants from '../constants/fs'
+import * as FsGen from './fs-gen'
+import * as I from 'immutable'
+import * as RPCTypes from '../constants/types/rpc-gen'
+import * as Saga from '../util/saga'
+import * as Types from '../constants/types/fs'
+import * as FsPath from '../constants/fs-path'
+
+function* folderList(action: FsGen.FolderListLoadPayload): Saga.SagaGenerator<any, any> {
+  if (action.payload.path === '/keybase') {
```
Comment (chrisnojima) [inline shared/actions/fs.js]: Why is this necessary to cancel

File: shared/actions/fs.js [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    return
+  }
+
+  const opID = Constants.makeUUID()
+  const rootPath = action.payload.path
+
+  yield Saga.call(RPCTypes.SimpleFSSimpleFSListRpcPromise, {
+    opID,
+    path: {
+      PathType: 1,
```
Comment (chrisnojima) [inline shared/actions/fs.js]: Isn't there a const for this that's generated

**Comment by chrisnojima** [comment]
I don't understand the argument for this path function from node. The data comes from us and is clean right so why isn't it just a simple join with /

**Comment by chrisnojima** [comment]
I'd prefer to make those constants than pull in this code

**Comment by songgao** [comment]
@chrisnojima We'll be supporting pasting in paths in M2 or something right? I was thinking that path will need to be cleaned.

**Comment by songgao** [comment]
> I'd prefer to make those constants than pull in this code

Hmm sorry I don't fully understand. When navigating to `/keybase`, it'd still trigger the RPC right?

I guess alternatively I could also make the SimpleFS change first before merging this PR so this code never goes in master.

**Comment by chrisnojima** [comment]
Let's not add complexity for things that aren't in m1. This is literally the first pr to actually hook up the data. Let's not focus on future features like pasting paths

**Comment by songgao** [comment]
OK sounds good; will fix!

**Comment by songgao** [comment]
@chrisnojima please take another look; thanks!!

Note that the FS tab will appear broken until https://github.com/keybase/kbfs/pull/1480 is in.

**Comment by songgao** [comment]
Thanks! Will hold off merging until KBFS side is in.
